p2p: unbroadcast followups: rpcs, nLastResend, mempool sanity check

glozow opened the pull request:
Followup to #18038 by amitiuttarwar which introduces the unbroadcast set: "a mechanism for the mempool to track locally submitted transactions" and decreases the frequency of rebroadcast from 10-15 minutes to 12-36 hours. 

This PR addresses some of the outstanding TODOs building on top of it:
- remove `nLastResend` logic, which is used to ensure rebroadcast doesn't happen again if no new block has been mined (makes sense in 10-15 min period, but not necessary for 12-36 hour period). (#18038 [comment](https://github.com/bitcoin/bitcoin/pull/18038#discussion_r416826914))
- expose unbroadcast info via RPCs, for more informative queries and testing (#18038 [comment](https://github.com/bitcoin/bitcoin/pull/18038#discussion_r416837980))
- add sanity check to verify unbroadcast transactions still exist in mempool before rebroadcasting (#18038 [comment](https://github.com/bitcoin/bitcoin/pull/18038#discussion_r416861609))

glozow committed:
[rpc] add unbroadcast info to mempool entries and getmempoolinfo

- expose info about number of txns in unbroadcast set and whether a mempool entry's tx has passed initial broadcast
- makes rpcs more informative and allows for more explicit testing, eg tracking if tx is in unbroadcast set
before and after originating node connects to peers (adds this in mempool_unbroadcast.py)
- adds mempool method IsUnbroadcastTx to query for tx inclusion in  mempool's unbroadcast set

glozow committed:
[mempool] sanity check that all unbroadcast txns are in mempool

- before reattempting broadcast for unbroadcast txns, check they are in mempool and remove if not
- this protects from memory leaks and network spam just in case unbroadcast set (incorrectly) has extra txns
- check that tx is in mempool before adding to unbroadcast set to try to prevent this from happening

glozow committed:
[wallet] remove nLastResend logic

remove nLastResend because it's unnecessary now that rebroadcasts always happen at least 12 hours later

fanquake commented:
My assumption is that those TODOs would be taken care of as part of #18807. Is that correct @amitiuttarwar ?

amitiuttarwar commented:
@fanquake sorry for the confusion 😛
@gzhao408 offered to implement these follow ups, so I asked her to open a PR so I & others could review. I was thinking (now that this PR is open), I'd update #18807 to let reviewers know, & focus that PR on updates to release notes and tests. 
does that seem reasonable?

fanquake commented:
> does that seem reasonable?

@amitiuttarwar Yea of course. Just wanted to clarify that you weren't both working on the same changes in parallel.

amitiuttarwar commented:
@fanquake thanks!

Concept ACK :) will review soon

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -815,7 +815,12 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
     for (const uint256& txid : unbroadcast_txids) {
-        RelayTransaction(txid, *connman);
+        // Sanity check: all unbroadcast txns should exist in the mempool
+        if (!m_mempool.exists(txid)) {
```
```suggestion
        if (m_mempool.exists(txid)) {
```

style-nit: (feel free to ignore)

When an `if` has an else branch, I generally recommend not using inversion in the condition, just to avoid double negation, which can make code harder to read. Obviously not an issue here, so feel free to ignore.

However, while this fights the symptom (unbroadcast txid exists), it doesn't prevent it from happening. Also, `RelayTransaction` is hopefully already a noop if the txid does not exist.

What do you think about adding this check to the mempool member function that adds the unbroadcast transactions to the set? This way, the error will be caught as early as possible.

gzhao408 replied:
Good point about inversion on an if+else, I'll rearrange to make it more clear.

Yeah, I think it'd make sense to add this check to `AddUnbroadcastTx` ([here](https://github.com/bitcoin/bitcoin/blob/7bcc42b4035b878719d13201286e322989b415c5/src/txmempool.h#L705)).
Apart from making sure we only add txids that exist in the mempool to unbroadcast, I think this sanity check helps if a tx was removed from the mempool but somehow stayed in the unbroadcast set - which shouldn't happen, but just in case?

MarcoFalke replied:
Jup, better to do the check in two places than in one. This needs a mempool lookup, but the cost doesn't seem too much, as this is only done every 10 minutes or when a new transaction is added to the mempool anyway.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1389,6 +1391,7 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("maxmempool", (int64_t) maxmempool);
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
+    ret.pushKV("unbroadcastsize", (int64_t)pool.GetUnbroadcastTxs().size());
```
style-nit: instead of a "blind cast" I slightly prefer the one that enables compile time checks and fails compilation in case it should be impossible to cast without loss of precision.

```suggestion
    ret.pushKV("unbroadcastsize", int64_t{pool.GetUnbroadcastTxs().size())};
```

gzhao408 replied:
That makes sense. The only issue is it wouldn't be consistent with all the other casts throughout the file. I'm not sure what's the best practice: let this one be inconsistent, change all of them in this PR, or change all of them in a followup PR?

MarcoFalke replied:
The guideline is that new code should follow best practices, old code can stay as is.

amitiuttarwar replied:
it's tough to aim for consistency in bitcoin bc the project values careful review for any changes (thus refactor PRs are generally only considered if there's a feature being built on top) so generally, we try to improve over time as we touch lines of code. 

another example you'll see is mismatched naming conventions. code that's been touched recently adheres to the best practices, but code that's been around for longer still follows old conventions 

(tldr- improve what you touch over time instead of aiming for consistency)

gzhao408 replied:
Makes sense, thank you!

maflcko reviewed (approved):
Concept ACK, but it looks like the `mempool_packages` has a hardcoded mempool entry in the test, so that fails because the hardcoded entry is missing the `unbraodcast` key.

mzumsande reviewed (commented):
Concept ACK. 

I think that `mempool_packages.py` just needs to wait a bit so that the unbroadcast status of txns does not change in between the `getrawmempool` and `getmempoolentry` calls because of delays in the p2p.

Also, it's confusing that #18807 has (almost) the same name as this PR, so maybe change the title?

naumenkogs commented:
Concept ACK. Code looks good, will do ACK when @MarcoFalke comments are addressed.

glozow committed:
[test] wait for inital broadcast before comparing mempool entries

- mempool entry 'unbroadcast' field changes when tx passes initial broadcast (receive getdata),
so anytime you compare mempool entries as a whole, you must wait for all broadcasts to complete
('unbroadcast' = False) otherwise the state may change in between calls
- update P2PTxInvStore to send msg_getdata for invs and add functionality to wait for a list
of txids to complete initial broadcast
- make mempool_packages.py wait because it compares entries using getrawmempool and
getmempoolentry

gzhao408 commented:
Sorry for all the force pushes 😅. Addressed MarcoFalke's comments and the `mempool_packages.py` error. As mzumsande said, it was due to the `unbroadcast` value changing in between calls (i.e. a transaction completed initial broadcast while the test was running, so a subsequent call with `verbose=True` was inconsistent with before. 

With `unbroadcast` as one of the mempool entry fields, it's probably not a good idea to compare mempool entries as a whole, ie `assert_equal(mempool0[tx], mempool1[tx]` because `unbroadcast` might be different. Most tests are more specific than this, so they are ok. I fixed `mempool_packages.py` by having the node wait for `len(mempool)` invs, so all `unbroadcast`s should be False.

mzumsande commented:
I meant an earlier spot in my comment: The `assert_equal(entry, mempool[x])` in L92 can fail because 
`mempool = self.nodes[0].getrawmempool(True)` and 
`entry = self.nodes[0].getmempoolentry(x)` 
can have inconsistent unbroadcast flags if we haven't waited before calling `getrawmempool` the first time.

gzhao408 commented:
> I meant an earlier spot in my comment: The `assert_equal(entry, mempool[x])` in L92 
@mzumsande OMG I am so sorry 🤦‍♀️ that should have been obvious. Just fixed it (I hope), waiting to see what travis says.

MarcoFalke commented on test/functional/mempool_packages.py:
```diff
@@ -72,6 +74,10 @@ def run_test(self):
             value = sent_value
             chain.append(txid)
 
+        # Wait until mempool transactions have passed initial broadcast (got an inv for each one)
```
The condition for initial broadcast is not that an inv was sent, but that a getdata was received

I think you can simply replace this with a `self.sync_mempools()` and remove the newly added p2p connection

gzhao408 replied:
> I think you can simply replace this with a self.sync_mempools() and remove the newly added p2p connection

I've tried this, but it times out: `AssertionError: Mempool sync timed out after 60s`

> The condition for initial broadcast is not that an inv was sent, but that a getdata was received

Oh this is my misunderstanding. I can try to have a `wait_until` that counts the `msg_getdata`s received by `nodes[0]`?

MarcoFalke replied:
Oh, I missed that the second node has a stricter `limitancestorcount` set.

In that case your suggestion to wait for the getdata makes most sense, I guess.

So 

* register callback in mininode to send getdata when inv arrives
* wait for invs
* sync_with_ping (to flush all buffers)

should work

gzhao408 replied:
Thank you, this is super helpful! 🙏

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1389,7 +1391,7 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
-
+    ret.pushKV("unbroadcastsize", static_cast<int64_t>(pool.GetUnbroadcastTxs().size()));
```
```suggestion
    ret.pushKV("unbroadcastsize", int64_t{pool.GetUnbroadcastTxs().size())};
```

This does the same thing, but has additional compile time checks that fail compilation when the conversion comes with a loss of precision.

gzhao408 replied:
Really sorry about this - I got a compilation error: 
```
rpc/blockchain.cpp:1394:43: error: non-constant-expression cannot be narrowed from type 'std::__1::set<uint256, std::__1::less<uint256>,
      std::__1::allocator<uint256> >::size_type' (aka 'unsigned long') to 'int64_t' (aka 'long long') in initializer list [-Wc++11-narrowing]
    ret.pushKV("unbroadcastsize", int64_t{pool.GetUnbroadcastTxs().size()});
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rpc/blockchain.cpp:1394:43: note: insert an explicit cast to silence this issue
    ret.pushKV("unbroadcastsize", int64_t{pool.GetUnbroadcastTxs().size()});
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                          static_cast<int64_t>(          )
```
So I followed the suggestion to fix it. I'm not sure why it didn't work for me.

MarcoFalke replied:
Oh, I guess you can use this signature `bool pushKV(const std::string&, uint64_t)`. It should already be selected by default, so I don't see why we want to cast in the first place.

I think you can either:

* Remove the cast
* Select the signature for `uint64_t`

MarcoFalke commented on test/functional/mempool_packages.py:
```diff
@@ -58,6 +59,7 @@ def chain_transaction(self, node, parent_txid, vout, value, fee, num_outputs):
 
     def run_test(self):
         # Mine some blocks and have them mature.
+        self.nodes[0].add_p2p_connection(P2PTxInvStore()) # keep track of invs
```
You still need to provide the callback. The default one does not reply with a getdata. See

```py
class P2PTxInvStore(P2PInterface):
    """A P2PInterface which stores a count of how many times each txid has been announced."""
    def __init__(self):
        super().__init__()
        self.tx_invs_received = defaultdict(int)

    def on_inv(self, message):
        # Store how many times invs have been received for each tx.
        for i in message.inv:
            if i.type == MSG_TX:
                # save txid
                self.tx_invs_received[i.hash] += 1

MarcoFalke replied:
You can run `git grep -W 'def on_inv'` to find examples on how to do that.

MarcoFalke replied:
Hmm. Looks like the tests passed on travis. I am still unsure why.

gzhao408 replied:
Gah, I made a super dumb assumption that it would automatically call `P2PInterface`'s `on_inv` (which responds with the `getdata`) since it inherits from it. I'm working on that right now. Not sure why travis passed haha

MarcoFalke replied:
I reset travis, so that it properly shows failure. :sweat_smile:

gzhao408 replied:
I spoke with @amitiuttarwar and decided to just have `P2PTxInvStore` call `super().on_inv` since she's planning on doing it anyway (in [#18807](https://github.com/bitcoin/bitcoin/pull/18807/files#diff-a77567bbc57dc1bbc033560e5cdd3b7dR647)). We've agreed to resolve any conflicts that happen depending on which gets merged first.

MarcoFalke commented on test/functional/test_framework/mininode.py:
```diff
@@ -647,3 +648,11 @@ def on_inv(self, message):
+    def wait_for_broadcast(self, num_txns):
+        # Waits for num_txns transactions to complete initial broadcast.
+        # Wait until num_txns invs have been received (and getdatas sent).
+        wait_until(lambda: len(self.tx_invs_received.keys()) == num_txns)
```
```suggestion
        self.wait_until(lambda: set(self.get_invs()) == set(txns))
```

* self.wait_until because we need the scaling factor
* self.get_inv because we need to take the lock
* set because we want to check the txids

gzhao408 replied:
This makes sense. 

It seems like a `P2PInterface`s don't have a `self.wait_until` function. I get a `object has no attribute 'wait_until'` error. The classes in `mininode.py` use `wait_until` with the `mininode_lock`.  I'm not familiar with what you mean by scaling factor, but is there something else I could do?

MarcoFalke replied:
Are you sure you are on current master? I am seeing this function locally:

```
$ git grep 'def wait_until'
test/functional/test_framework/mininode.py:    def wait_until(self, test_function, timeout):

maflcko reviewed (approved):
Almost ACK

amitiuttarwar commented on src/txmempool.h:
```diff
@@ -704,7 +704,10 @@ class CTxMemPool
     void AddUnbroadcastTx(const uint256& txid) {
         LOCK(cs);
-        m_unbroadcast_txids.insert(txid);
+        /** Sanity Check: the transaction should also be in the mempool */
```
hm, is there a reason this is a [doxygen comment](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments)? I'm not sure what would happen, but I suspect you want normal `//` notation here

jonatack replied:
> I suspect you want normal `//` notation here

Agree, perhaps can be touched up in #18807

amitiuttarwar commented on test/functional/test_framework/mininode.py:
```diff
@@ -654,3 +655,11 @@ def on_inv(self, message):
         with mininode_lock:
             return list(self.tx_invs_received.keys())
+
+    def wait_for_broadcast(self, txns):
+        # Waits for the txns (txids) to complete initial broadcast.
+        # Wait until invs have been received (and getdatas sent) for each txid.
+        self.wait_until(lambda: set(self.get_invs()) == set([int(tx, 16) for tx in txns]), 60)
+        # Flush messages and wait for the getdatas to be processed
+        # The mempool should mark unbroadcast=False for these transactions.
+        self.sync_with_ping()
```
nit: pull out the description of the function from the implementation & description for readability 

```suggestion
    # Waits for the txns (txids) to complete initial broadcast.
    # The mempool should mark unbroadcast=False for these transactions.
    def wait_for_broadcast(self, txns):
        # Wait until invs have been received (and getdatas sent) for each txid.
        self.wait_until(lambda: set(self.get_invs()) == set([int(tx, 16) for tx in txns]), 60)
        # Flush messages and wait for the getdatas to be processed
        self.sync_with_ping()
```

gzhao408 replied:
Good point, gonna follow [this PEP-8 guideline](https://www.python.org/dev/peps/pep-0008/#documentation-strings)

amitiuttarwar reviewed (commented):
overall, the code looks good! the approaches are simple & make sense. solid test coverage. I also appreciate the thoughtful documentation :) thank you for taking this on! 

I left a bunch of nits and thoughts to consider. Other than removing the `const` to get rid of the build warning, feel free to take or leave whatever.

Just to note- this code adds unbroadcast info to 5 different RPC methods. Of these 5, the unbroadcast info of 2 are explicitly tested from the functional tests (`getmempoolinfo`, `getrawmempool`). But the others are implicitly tested because its a shared data structure that serves the data (`getmempoolancestors`, `getmempooldescendants`, `getmempoolentry`). That coverage seems sufficient to me.

I like the `wait_for_broadcast` method & how it encapsulates the synchronization required to ensure a txn is removed from the unbroadcast set. have you checked if there are other rpc call sites in the tests that might benefit (aka help prevent weird race conditions)? I’ll double check this for my next round of review. 

thanks!

amitiuttarwar commented:
also I realized, since this updates some RPC endpoints, should there be release notes added? @MarcoFalke

MarcoFalke commented:
Sure, but I assumed this would be taken care of in the doc pr #18807

gzhao408 commented on test/functional/mempool_packages.py:
```diff
@@ -140,6 +146,7 @@ def run_test(self):
                     assert_equal(ainfo['depends'], [])
 
 
+
```
😅 I'll delete this

MarcoFalke commented on test/functional/test_framework/mininode.py:
```diff
@@ -654,3 +655,11 @@ def on_inv(self, message):
+    def wait_for_broadcast(self, txns):
+        # Waits for the txns (txids) to complete initial broadcast.
+        # Wait until invs have been received (and getdatas sent) for each txid.
+        self.wait_until(lambda: set(self.get_invs()) == set([int(tx, 16) for tx in txns]), 60)
```
```suggestion
        self.wait_until(lambda: set(self.get_invs()) == set([int(tx, 16) for tx in txns]), timeout)
```

I think it makes sense to define a default value of `timeout=60` for the wait_for_broadcast member function. This way, callers can decide to overwrite it if needed.

gzhao408 commented on src/txmempool.h:
```diff
@@ -716,6 +719,12 @@ class CTxMemPool
     }
 
+    /** Returns if a txid is in the unbroadcast set */
+    const bool InUnbroadcast(const uint256& txid) const {
```
Ok after over-thinking this for a few hours 😂 I'm going to change it to `IsUnbroadcastTx` which looks pretty similar to the others and hopefully feels natural.
I definitely think `entry.IsUnbroadcast()` would be ideal if it were a mempool entry method (I didn't do it that way because `m_unbroadcast_txids` belongs to the mempool). I just don't want it to sound like I'm saying the mempool is unbroadcast.

MarcoFalke commented:
Review ACK 651f1d816f054cb9c637f8a99c9360bba381ef58

naumenkogs reviewed (approved):
Code review ACK 651f1d816f054cb9c637f8a99c9360bba381ef58

amitiuttarwar commented:
ACK 651f1d816f054cb9c637f8a99c9360bba381ef58 🎉 

double checked functional tests for possible races around returning `unbroadcast` from the various RPCs. looks like `mempool_packages` is the only one comparing all the fields of the returned object, so your changes have them covered.

jonatack commented on src/txmempool.h:
```diff
@@ -716,6 +719,12 @@ class CTxMemPool
         return m_unbroadcast_txids;
     }
 
+    // Returns if a txid is in the unbroadcast set
```
This can be doxygen like the neighbouring functions. Perhaps touch up in #18807.

amitiuttarwar replied:
done

jonatack reviewed (commented):
Posthumous code review ACK 651f1d816f054cb9c637f8a99c9360bba381ef58

harding commented on src/rpc/blockchain.cpp:
```diff
@@ -399,6 +399,7 @@ static std::vector<RPCResult> MempoolEntryDescription() { return {
     RPCResult{RPCResult::Type::ARR, "spentby", "unconfirmed transactions spending outputs from this transaction",
         {RPCResult{RPCResult::Type::STR_HEX, "transactionid", "child transaction id"}}},
     RPCResult{RPCResult::Type::BOOL, "bip125-replaceable", "Whether this transaction could be replaced due to BIP125 (replace-by-fee)"},
+    RPCResult{RPCResult::Type::BOOL, "unbroadcast", "Whether this transaction is currently unbroadcast (initial broadcast not yet confirmed)"},
```
Suggested rephrasing: s/confirmed/acknowledged/.  I think "confirmed" might confuse the user into thinking we're referring to the transaction being included in a block.  Perhaps "acknowledged" would be less confusing, or if a few more words are ok, "(initial broadcast not yet acknowledged by any peers)".  Maybe @amitiuttarwar can squeeze this in #18807 (if not, I can open a separate PR).

amitiuttarwar replied:
great suggestion. adding to #18807

amitiuttarwar replied:
done
